File: src/offline/Queue.fs
module Offline.Queue

open Fable.Core
open Fable.Core.JsInterop
open Offline.Types

/// Database name and version
let private dbName = "rollbook-offline"
let private dbVersion = 2
let private storeName = "queue"

/// Import idb library
[<Import("openDB", from="idb")>]
let private openDB: string -> int -> obj -> JS.Promise<obj> = jsNative

/// Get or create the database (v2: clears v1 queue on upgrade)
let private getDb () : JS.Promise<obj> =
    let upgradeConfig =
        emitJsExpr storeName """({
            upgrade(db, oldVersion, _newVersion, transaction) {
                if (!db.objectStoreNames.contains($0)) {
                    db.createObjectStore($0, { keyPath: 'id', autoIncrement: true });
                } else if (oldVersion < 2) {
                    transaction.objectStore($0).clear();
                    console.log('Cleared v1 queue during upgrade to v2');
                }
            }
        })"""
    openDB dbName dbVersion upgradeConfig

/// Enqueue a workout operation for offline sync
let enqueue (operationType: OperationType) (userId: string) (workoutDate: string) : JS.Promise<QueueResult> =
    promise {
        try
            let! db = getDb ()
            let operation = {
                id = None
                operationType =
                    match operationType with
                    | CreateWorkout -> "CreateWorkout"
                    | DeleteWorkout -> "DeleteWorkout"
                recordId = None
                userId = userId
                workoutDate = workoutDate
                recordType = "workout"
                textContent = None
                photoUrl = None
                timestamp = JS.Constructors.Date.now()
                retryCount = 0
            }
            let! id = db?add(storeName, operation)
            return Queued (unbox<int> id)
        with exn ->
            return QueueError exn.Message
    }

/// Get all pending operations
let getAllPending () : JS.Promise<QueuedOperation array> =
    promise {
        try
            let! db = getDb ()
            let! items = db?getAll(storeName)
            return unbox<QueuedOperation array> items
        with _ ->
            return [||]
    }

/// Remove an operation from the queue (after successful sync)
let dequeue (operationId: int) : JS.Promise<bool> =
    promise {
        try
            let! db = getDb ()
            do! db?delete(storeName, operationId)
            return true
        with _ ->
            return false
    }

/// Update retry count for a failed operation
let incrementRetry (operationId: int) : JS.Promise<bool> =
    promise {
        try
            let! db = getDb ()
            let! operation = db?get(storeName, operationId)
            match box operation with
            | null -> return false
            | _ ->
                let updated =
                    createObj [
                        "id" ==> operationId
                        "operationType" ==> operation?operationType
                        "recordId" ==> operation?recordId
                        "userId" ==> operation?userId
                        "workoutDate" ==> operation?workoutDate
                        "recordType" ==> operation?recordType
                        "textContent" ==> operation?textContent
                        "photoUrl" ==> operation?photoUrl
                        "timestamp" ==> operation?timestamp
                        "retryCount" ==> (unbox<int> operation?retryCount + 1)
                    ]
                do! db?put(storeName, updated)
                return true
        with _ ->
            return false
    }

/// Get count of pending operations
let getPendingCount () : JS.Promise<int> =
    promise {
        try
            let! db = getDb ()
            let! count = db?count(storeName)
            return unbox<int> count
        with _ ->
            return 0
    }

/// Clear all queued operations (use with caution)
let clear () : JS.Promise<bool> =
    promise {
        try
            let! db = getDb ()
            do! db?clear(storeName)
            return true
        with _ ->
            return false
    }

File: src/offline/Queue.js
import { openDB } from "idb";
import { PromiseBuilder__Delay_62FBFDE1, PromiseBuilder__Run_212F1D4B } from "../fable_modules/Fable.Promise.3.2.0/Promise.fs.js";
import { promise } from "../fable_modules/Fable.Promise.3.2.0/PromiseImpl.fs.js";
import { QueueResult, QueuedOperation } from "./Types.js";
import { defaultOf, equals } from "../fable_modules/fable-library-js.4.28.0/Util.js";

const dbName = "rollbook-offline";

const dbVersion = 1;

const storeName = "queue";

function getDb() {
    const upgradeConfig = {
        upgrade: (db) => {
            if (!(db.objectStoreNames.contains(storeName))) {
                db.createObjectStore(storeName, {
                    keyPath: "id",
                    autoIncrement: true,
                });
            }
        },
    };
    return openDB(dbName)(dbVersion)(upgradeConfig);
}

/**
 * Enqueue a workout operation for offline sync
 */
export function enqueue(operationType, userId, workoutDate) {
    return PromiseBuilder__Run_212F1D4B(promise, PromiseBuilder__Delay_62FBFDE1(promise, () => (PromiseBuilder__Delay_62FBFDE1(promise, () => (getDb().then((_arg) => {
        const db = _arg;
        const operation = new QueuedOperation(undefined, (operationType.tag === 1) ? "DeleteWorkout" : "CreateWorkout", userId, workoutDate, Date.now(), 0);
        return (db.add(storeName, operation)).then((_arg_1) => {
            const id = _arg_1;
            return Promise.resolve(new QueueResult(0, [id]));
        });
    }))).catch((_arg_2) => {
        const exn = _arg_2;
        return Promise.resolve(new QueueResult(1, [exn.message]));
    }))));
}

/**
 * Get all pending operations
 */
export function getAllPending() {
    return PromiseBuilder__Run_212F1D4B(promise, PromiseBuilder__Delay_62FBFDE1(promise, () => (PromiseBuilder__Delay_62FBFDE1(promise, () => (getDb().then((_arg) => {
        const db = _arg;
        return (db.getAll(storeName)).then((_arg_1) => {
            const items = _arg_1;
            return Promise.resolve(items);
        });
    }))).catch((_arg_2) => (Promise.resolve([]))))));
}

/**
 * Remove an operation from the queue (after successful sync)
 */
export function dequeue(operationId) {
    return PromiseBuilder__Run_212F1D4B(promise, PromiseBuilder__Delay_62FBFDE1(promise, () => (PromiseBuilder__Delay_62FBFDE1(promise, () => (getDb().then((_arg) => {
        const db = _arg;
        return (db.delete(storeName, operationId)).then(() => (Promise.resolve(true)));
    }))).catch((_arg_2) => (Promise.resolve(false))))));
}

/**
 * Update retry count for a failed operation
 */
export function incrementRetry(operationId) {
    return PromiseBuilder__Run_212F1D4B(promise, PromiseBuilder__Delay_62FBFDE1(promise, () => (PromiseBuilder__Delay_62FBFDE1(promise, () => (getDb().then((_arg) => {
        const db = _arg;
        return (db.get(storeName, operationId)).then((_arg_1) => {
            const operation = _arg_1;
            if (equals(operation, defaultOf())) {
                return Promise.resolve(false);
            }
            else {
                const updated = {
                    id: operationId,
                    operationType: operation.operationType,
                    userId: operation.userId,
                    workoutDate: operation.workoutDate,
                    timestamp: operation.timestamp,
                    retryCount: operation.retryCount + 1,
                };
                return (db.put(storeName, updated)).then(() => (Promise.resolve(true)));
            }
        });
    }))).catch((_arg_3) => (Promise.resolve(false))))));
}

/**
 * Get count of pending operations
 */
export function getPendingCount() {
    return PromiseBuilder__Run_212F1D4B(promise, PromiseBuilder__Delay_62FBFDE1(promise, () => (PromiseBuilder__Delay_62FBFDE1(promise, () => (getDb().then((_arg) => {
        const db = _arg;
        return (db.count(storeName)).then((_arg_1) => {
            const count = _arg_1;
            return Promise.resolve(count);
        });
    }))).catch((_arg_2) => (Promise.resolve(0))))));
}

/**
 * Clear all queued operations (use with caution)
 */
export function clear() {
    return PromiseBuilder__Run_212F1D4B(promise, PromiseBuilder__Delay_62FBFDE1(promise, () => (PromiseBuilder__Delay_62FBFDE1(promise, () => (getDb().then((_arg) => {
        const db = _arg;
        return (db.clear(storeName)).then(() => (Promise.resolve(true)));
    }))).catch((_arg_2) => (Promise.resolve(false))))));
}

